File: deployment/tests/six_job_test.py
import os
import sys
import uuid

from jobmon.client.api import Tool


thisdir = os.path.dirname(os.path.realpath(__file__))


def six_job_test(cluster_name: str):
    """
    Creates and runs one workflow with six jobs. Used to 'smoke test' a
    new deployment of jobmon.
    """
    # First Tier
    # Deliberately put in on the long queue with max runtime > 1 day
    tool = Tool(name="Iguanodon alpha testing - Slurm")
    long_q_template = tool.get_task_template(
        template_name='long_queue_test',
        command_template="{command}",
        node_args=['command']
    )
    normal_q_template = tool.get_task_template(
        template_name='normal_queue_test',
        command_template="{command}",
        node_args=['command']
    )
    tool.set_default_compute_resources_from_yaml(
        default_cluster_name=cluster_name,
        yaml_file=os.path.join(thisdir, "six_job_test_resouces.yaml"),
        set_task_templates=True,
        ignore_missing_keys=True
    )

    t1 = long_q_template.create_task(
        name='t1',
        command="sleep 10"
    )

    # Second Tier, both depend on first tier
    t2 = normal_q_template.create_task(
        name='t2',
        command="sleep 20",
        upstream_tasks=[t1]
    )

    t3 = normal_q_template.create_task(
        name='t3',
        command="sleep 25",
        upstream_tasks=[t1]
    )

    # Third Tier, cross product dependency on second tier
    t4 = normal_q_template.create_task(
        name='t4',
        command="sleep 17",
        upstream_tasks=[t2, t3]
    )

    t5 = normal_q_template.create_task(
        name='t5',
        command="sleep 13",
        upstream_tasks=[t2, t3]
    )

    # Fourth Tier, ties it all back together
    t6 = normal_q_template.create_task(
        name='t6',
        command="sleep 19",
        upstream_tasks=[t4, t5]
    )

    wf = tool.create_workflow(
        workflow_args="six-job-test-{}".format(uuid.uuid4()),
        name='six_job_test')
    wf.add_tasks([t1, t2, t3, t4, t5, t6])
    print("Running the workflow, about 70 seconds minimum")
    wfr_status = wf.run()
    if wfr_status == 'D':
        print("All good, dancing pixies.")
    else:
        raise ValueError(f"Workflow should be successful, not state {wfr_status}")


if __name__ == "__main__":
    cluster_name = sys.argv[1]
    six_job_test(cluster_name)

File: deployment/tests/six_job_test.r
library(argparse)
parser <- ArgumentParser()
parser$add_argument("--python-path",
                    help = "Path to python executable")
parser$add_argument("--jobmonr-loc",
                    help = "The location of jobmonr")

args <- parser$parse_args()
python_path <- args$python_path
jobmonr_loc <- args$jobmonr_loc

Sys.setenv("RETICULATE_PYTHON" = python_path)  # Set the Python interpreter path

library(devtools)
# install from source an in-memory package
devtools::load_all(path=jobmonr_loc)


# Bind a workflow to the tool
my_tool <- tool()
wf <- workflow(tool=my_tool, name=paste0("six-job-test-", runif(1)))

# Set the executor
wf <- set_executor(wf, executor_class='SGEExecutor', project="ihme_general")

# Define executor parameters for our tasks
params <- executor_parameters(num_cores=1, m_mem_free="1G", queue='all.q', max_runtime_seconds=100)

# Define task template
sleep_template <- task_template(tool=my_tool,
                                template_name='sleep_template',
                                command_template='sleep {sleep_time}',
                                node_args=list('sleep_time'))


# Create tasks
t1 <- task(task_template=sleep_template,
           executor_parameters=params,
           name='sleep_1',
           sleep_time="1")

# Second Tier, both depend on first tier
t2 <- task(task_template=sleep_template,
           executor_parameters=params,
           name='sleep_2',
           sleep_time="20",
           upstream_tasks = list(t1))
t3 <- task(task_template=sleep_template,
           executor_parameters=params,
           name='sleep_3',
           sleep_time="25",
           upstream_tasks = list(t1))

# Third Tier, cross product dependency on second tier
t4 <- task(task_template=sleep_template,
           executor_parameters=params,
           name='sleep_4',
           sleep_time="10",
           upstream_tasks = list(t2, t3))
t5 <- task(task_template=sleep_template,
           executor_parameters=params,
           name='sleep_5',
           sleep_time="13",
           upstream_tasks = list(t2, t3))

# Fourth Tier, ties it all back together
t6 <- task(task_template=sleep_template,
           executor_parameters=params,
           name='sleep_6',
           sleep_time="19",
           upstream_tasks = list(t4, t5))


# Add tasks to the workflow
wf <- add_tasks(wf, list(t1, t2, t3, t4, t5, t6))

# Run it
wfr <- run(workflow=wf, resume=F, seconds_until_timeout=600)
